pkg/rehearse/jobs.go

```diff
@@ -110,45 +110,76 @@ func makeRehearsalPresubmit(source *prowconfig.Presubmit, repo string, prNumber
 	return &rehearsal, nil
 }
 
-func filterJobs(changedPresubmits map[string][]prowconfig.Presubmit, allowVolumes bool, logger logrus.FieldLogger) config.Presubmits {
-	ret := config.Presubmits{}
+func makeRehearsalPeriodic(source *prowconfig.Periodic, prNumber int) (prowconfig.Periodic, error) {
+	var rehearsal prowconfig.Periodic
+	deepcopy.Copy(&rehearsal, source)
+
+	rehearsal.Name = fmt.Sprintf("rehearse-%d-%s", prNumber, source.Name)
+	if rehearsal.Labels == nil {
+		rehearsal.Labels = make(map[string]string, 1)
+	}
+	rehearsal.Labels[rehearseLabel] = strconv.Itoa(prNumber)
+
+	return rehearsal, nil
+}
+
+func filterPresubmits(changedPresubmits map[string][]prowconfig.Presubmit, allowVolumes bool, logger logrus.FieldLogger) config.Presubmits {
+	presubmits := config.Presubmits{}
 	for repo, jobs := range changedPresubmits {
 		for _, job := range jobs {
 			jobLogger := logger.WithFields(logrus.Fields{"repo": repo, "job": job.Name})
-			if err := filterJob(&job, allowVolumes); err != nil {
+			if len(job.Branches) == 0 {
+				jobLogger.Warn("cannot rehearse jobs with no branches")
+				continue
+			}
+
+			if len(job.Branches) != 1 {
+				jobLogger.Warn("cannot rehearse jobs that run over multiple branches")
+				continue
+			}
+
+			if err := filterJobSpec(job.Spec, allowVolumes, false); err != nil {
 				jobLogger.WithError(err).Warn("could not rehearse job")
 				continue
 			}
-			ret.Add(repo, job)
+			presubmits.Add(repo, job)
+		}
+	}
+	return presubmits
+}
+
+func filterPeriodics(changedPeriodics []prowconfig.Periodic, allowVolumes bool, logger logrus.FieldLogger) []prowconfig.Periodic {
+	var periodics []prowconfig.Periodic
+	for _, periodic := range changedPeriodics {
+		jobLogger := logger.WithField("job", periodic.Name)
+		if err := filterJobSpec(periodic.Spec, allowVolumes, true); err != nil {
+			jobLogger.WithError(err).Warn("could not rehearse job")
+			continue
 		}
+		periodics = append(periodics, periodic)
 	}
-	return ret
+	return periodics
 }
 
-func filterJob(source *prowconfig.Presubmit, allowVolumes bool) error {
+func filterJobSpec(spec *v1.PodSpec, allowVolumes, allowGitRef bool) error {
 	// there will always be exactly one container.
-	container := source.Spec.Containers[0]
+	container := spec.Containers[0]
 
 	if len(container.Command) != 1 || container.Command[0] != "ci-operator" {
 		return fmt.Errorf("cannot rehearse jobs that have Command different from simple 'ci-operator'")
 	}
 
-	for _, arg := range container.Args {
-		if strings.HasPrefix(arg, "--git-ref") || strings.HasPrefix(arg, "-git-ref") {
-			return fmt.Errorf("cannot rehearse jobs that call ci-operator with '--git-ref' arg")
+	if !allowGitRef {
+		for _, arg := range container.Args {
+			if strings.HasPrefix(arg, "--git-ref") || strings.HasPrefix(arg, "-git-ref") {
+				return fmt.Errorf("cannot rehearse jobs that call ci-operator with '--git-ref' arg")
```

Comment:
Why do we allow this for periodics?

Reply:
Most of our periodics, already have `git-ref`. I guess this is how our stakeholders wanted to configure. Also, the `extra_refs` used to be ignore by ci-operator but this have been fixed by now. Shall we force them do otherwise?

Reply:
meaning that we need to change all of our current periodics in release repo.

Reply:
Let's make them opt-in. This is why. Opt in and follow the rules if you want to be rehearsed.